tokio/src/sync/watch.rs

```diff
@@ -412,18 +418,32 @@ mod state {
         }
 
         /// Load the current value of the state.
-        pub(super) fn load(&self) -> StateSnapshot {
-            StateSnapshot(self.0.load(SeqCst))
+        pub(super) fn load(&self, ordering: Ordering) -> StateSnapshot {
+            StateSnapshot(self.0.load(ordering))
+        }
+
+        /// Load the current value of the state.
+        ///
+        /// The receiver side (read-only) uses `Acquire` ordering for a proper handover
+        /// with the sender side (single writer).
+        pub(super) fn load_receiver(&self) -> StateSnapshot {
+            StateSnapshot(self.0.load(Ordering::Acquire))
         }
 
         /// Increment the version counter.
         pub(super) fn increment_version(&self) {
-            self.0.fetch_add(STEP_SIZE, SeqCst);
+            // Use `Release` ordering to ensure that storing the version
+            // state is seen by the receiver side that uses `Acquire` for
+            // loading the state.
+            self.0.fetch_add(STEP_SIZE, Ordering::Release);
         }
 
         /// Set the closed bit in the state.
         pub(super) fn set_closed(&self) {
-            self.0.fetch_or(CLOSED_BIT, SeqCst);
+            // Use `Release` ordering to ensure that storing the version
+            // state is seen by the receiver side that uses `Acquire` for
+            // loading the state.
+            self.0.fetch_or(CLOSED_BIT, Ordering::Release);
```

Comment:
I agree that AcqRel is not needed for either type of store. The version increments all hold the write guard, and setting the closed bit requires mutable access to the sender since it's in the destructor. This ensures that we already have a happens-before relationship between them.